pkg/operator/ceph/csi/spec.go

```diff
@@ -662,3 +688,265 @@ func (r *ReconcileCSI) validateCSIVersion(ownerInfo *k8sutil.OwnerInfo) (*CephCS
+
+	multusHostSetup, err := templateToDaemonSet(csiMultusSetup, MultusDaemonsetTemplatePath, tp)
+	if err != nil {
+		return errors.Wrap(err, "failed to get daemonset template to set up the csi network")
+	}
+
+	// Applying affinity and toleration to multus daemonset.
+	// The multus network must be configured on every node that csi plugin pods run on.
+	pluginTolerations := getToleration(r.opConfig.Parameters, pluginTolerationsEnv, []corev1.Toleration{})
+	pluginNodeAffinity := getNodeAffinity(r.opConfig.Parameters, pluginNodeAffinityEnv, &corev1.NodeAffinity{})
+	applyToPodSpec(&multusHostSetup.Spec.Template.Spec, pluginNodeAffinity, pluginTolerations)
+
+	err = k8sutil.CreateDaemonSet(csiMultusSetup, r.opConfig.OperatorNamespace, r.context.Clientset, multusHostSetup)
+	if err != nil {
+		return errors.Wrap(err, "failed to start daemonset to set up the csi network")
+	}
+
+	// Wait for daemonset pods to come up.
+	// Will poll daemonset to ensure pods are available for interface migration.
+	err = wait.Poll(5*time.Second, 30*time.Second, func() (bool, error) {
+		daemonset, err := r.context.Clientset.AppsV1().DaemonSets(r.opConfig.OperatorNamespace).Get(r.opManagerContext, csiMultusDaemonsetName, metav1.GetOptions{})
+		if err != nil {
+			return false, errors.Wrap(err, "failed to get multus daemonset")
+		}
+
+		if daemonset.Status.NumberReady == daemonset.Status.DesiredNumberScheduled {
+			return true, nil
+		}
+		return false, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// The multus address range is used to find the multus interface for migration and cleanup.
+	nad, err := r.context.NetworkClient.NetworkAttachmentDefinitions(r.opConfig.OperatorNamespace).Get(r.opManagerContext, publicNetwork, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrap(err, "failed to get network attachment definition")
+	}
+	multusRange, err := multus.GetAddressRange(nad.Spec.Config)
+	if err != nil {
+		return errors.Wrap(err, "failed to get multus address range")
+	}
+
+	// Deploy jobs to move network interface from daemonset pods to host network r.opConfig.OperatorNamespace.
+	podList, err := r.context.Clientset.CoreV1().Pods(r.opConfig.OperatorNamespace).List(r.opManagerContext, metav1.ListOptions{
+		LabelSelector: "app=rook-ceph-multus",
+	})
+	if err != nil || len(podList.Items) < 1 {
+		return errors.Wrap(err, "failed to get multus daemonset pods")
+	}
+
+	var migrationError bool
+	ch := make(chan migrationJobReport)
+
+	for _, pod := range podList.Items {
+		tp.MultusNodeName = pod.Spec.NodeName
+		tp.MultusPodIP = pod.Status.PodIP
+
+		tp.MultusIP, tp.MultusLink, err = multus.GetMultusConf(pod, publicNetwork, nad.ObjectMeta.Namespace, multusRange)
+		if err != nil {
+			logger.Errorf("failed to get multus IP from pod: %v", err)
+			migrationError = true
+			continue
+		}
+
+		// Will deploy job and wait for completion before moving on to next one.
+		// If a job fails, the nodes that have been modified will be reverted.
```

Comment:
Requiring all-or-none tends to be brittle. If a single node is flaky, it could cause the whole cluster to be unusable. Instead, how about an approach where the operator keeps trying to configure failed nodes, but allows others to be usable? This is the advantage of continuous reconciling by operators.